File: rsa/rsa.js
var input = require('fs').readFileSync('/home/devp/Área de Trabalho/teste/inputs/rsa_in_1', 'utf8');
var [n, e ,c] = input.split('\n')[0].split(' ');

var x, y, d;

/**
 * calculatePhi.
 *
 *
 * @param {number} n value
 *
 * @returns {number} phi
 */
const calculatePhi = (n) => {
  let ret = parseInt(n, 10);
  for (let i = 2; i * i <= n; i++) {
    if (parseInt(n % i, 10) === 0) {
      while (parseInt(n % i, 10) === 0) {
        n = parseInt(n/i, 10);
      }
      ret -= parseInt(ret/i, 10);
    }
  }
  if (n > 1) {
    ret -= parseInt(ret/n, 10);
  }
  return ret;
};

/**
 * fastExp.
 *
 *
 * @param {number} p value
 * @param {number} q value
 * @param {number} m value
 *
 * @returns {number} fastExp
 */
const fastExp = (p, q, m) => {
  p = parseInt(p, 10);
  q = parseInt(q, 10);
  m = parseInt(m, 10);
  let r = parseInt(1, 10);
  while (q) {
    if (q & 1) {
      r = parseInt((r*p)%m, 10);
    }
    p = parseInt((p*p)%m, 10);
    q = parseInt(q >> 1, 10);
  }
  return r;
};

/**
 * extendedGCD.
 *
 *
 * @param {number} a value
 * @param {number} b value
 *
 */
const extendedGCD = (a, b) => {
  if(!b) {
    x = 1;
    y = 0;
    d = parseInt(a, 10);
    return;
  }
  extendedGCD(b, a%b);
  const auxX = parseInt(y, 10);
  const auxY = parseInt(x - (a/b) * y, 10);
  x = parseInt(auxX, 10);
  y = parseInt(auxY, 10);
};

/**
 * multiplicativeInverse.
 *
 *
 * @param {number} a value
 * @param {number} m value
 *
 */
const multiplicativeInverse = (a, m) => {
  a = parseInt(a, 10);
  m = parseInt(m, 10);
  extendedGCD(a, m);
  return parseInt((x%m+m)%m, 10);
};

(function () {
  let phi;
  n = parseInt(n, 10);
  e = parseInt(e, 10);
  c = parseInt(c, 10);
  phi = parseInt(calculatePhi(n), 10);
  let d = multiplicativeInverse(e, phi);
  console.log(fastExp(c, d, n));
})();

File: rsa/rsa.c
#include <stdio.h>

int x, y, d;
typedef long long ll;
int calculate_phi(int);
void extended_gcd(int, int);
ll fast_exp(ll, ll, ll);
int multiplicative_inverse(int, int);

int main(int argc, char **argv) {
    int n, e, c, phi;
    scanf("%d %d %d", &n, &e, &c);
    phi = calculate_phi(n);
    int inverted_mod = multiplicative_inverse(e, phi);
    printf("%lld\n", fast_exp(c, inverted_mod, n));

    return 0;
}

int calculate_phi(int n) {
    int result = n;
    for (int i = 2; i * i <= n; i++) {
        if(n % i == 0) {
            while(n % i == 0)
                n /= i;
            result -= result / i;
        }
    }
    if(n > 1)
        result -= result / n;

    return result;

}

ll fast_exp(ll p, ll q, ll m){
    ll r = 1;
    for (; q; q >>= 1) {
        if (q & 1)
            r = (r * p) % m;
        p = (p * p) % m;
    }
    return r;
}

void extended_gcd(int a, int b) {
    if (!b) {
        x = 1;
        y = 0;
        d = a;
        return;
    }
    extended_gcd(b, a % b);
    int x1 = y, y1 = x - (a / b) * y;
    x = x1, y = y1;
}

int multiplicative_inverse(int a, int m) {
    extended_gcd(a, m);
    return (x % m + m) % m;
}
